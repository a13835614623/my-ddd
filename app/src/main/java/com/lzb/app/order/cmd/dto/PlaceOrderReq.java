package com.lzb.app.order.cmd.dto;

import java.math.BigDecimal;

/**
 * 生单指令<br/>
 * Created on : 2023-09-06 13:00
 * @author lizebin
 */
public record PlaceOrderReq(String currency, BigDecimal exchangeRate, BigDecimal totalShouldPay,
                            BigDecimal totalActualPay, String email, String phoneNumber, String firstName,
                            String lastName, String addressLine1, String addressLine2, String country) {


}
