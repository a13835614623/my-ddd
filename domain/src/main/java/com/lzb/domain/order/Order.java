package com.lzb.domain.order;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Objects;

import cn.hutool.core.lang.UUID;
import com.lzb.domain.common.BaseAggregate;
import com.lzb.domain.order.event.OrderCanceledEvent;
import lombok.Builder;
import lombok.Getter;
import lombok.NonNull;
import lombok.experimental.SuperBuilder;
import lombok.extern.jackson.Jacksonized;
import lombok.extern.slf4j.Slf4j;

/**
 * 订单聚合根<br/>
 * 实体所有写的操作，必须通过聚合根调用，所以改成包可见(default)
 * 实体所有读操作，可以改成公共(public)
 * Created on : 2022-06-11 14:41
 *
 * @author lizebin
 */
@Slf4j
@Getter
// builder 反序列化
@SuperBuilder
@Jacksonized
public class Order extends BaseAggregate<Order> {

    @NonNull
    private String orderStatus;

    @NonNull
    private String currency;

    @NonNull
    private BigDecimal exchangeRate;

    @NonNull
    private BigDecimal totalShouldPay;

    @NonNull
    private BigDecimal totalActualPay;

    @NonNull
    private OrderAddress orderAddress;

    @Builder.Default
    private OrderDetails orderDetails = new OrderDetails(new ArrayList<>());

    public void updateTotalActualPay(BigDecimal totalActualPay) {
        this.totalActualPay = totalActualPay;
    }

    /**
     * 取消订单
     */
    public void cancel() {
        orderStatus = "CANCEL";
        orderDetails.forEach(OrderDetail::cancel);
        addEvent(OrderCanceledEvent.create(
                UUID.randomUUID().toString(),
                Objects.toString(id),
                Objects.toString(id),
                id));
    }
}
