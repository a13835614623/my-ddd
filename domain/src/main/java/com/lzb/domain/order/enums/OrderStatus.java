package com.lzb.domain.order.enums;

import com.lzb.component.utils.enums.EnumIntValue;
import com.lzb.component.utils.enums.EnumValue;
import lombok.AllArgsConstructor;
import lombok.Getter;

/**
 * 订单状态<br/>
 * Created on : 2023-09-04 10:00
 * @author lizebin
 */
@Getter
@AllArgsConstructor
public enum OrderStatus implements EnumValue<Integer> {

    WAIT_PAY(0, "待支付"), PAID(1, "已支付"), CANCELED(2, "已取消");

    private final Integer value;
    private final String desc;

}