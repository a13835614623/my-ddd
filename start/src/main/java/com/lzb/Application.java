package com.lzb;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.context.annotation.EnableAspectJAutoProxy;

@SpringBootApplication
// AopContext.currentProxy() 启用自身调用aop
@EnableAspectJAutoProxy(proxyTargetClass = true, exposeProxy = true)
public class Application {

	public static void main(String[] args) {
		SpringApplication.run(Application.class, args);
	}

}
