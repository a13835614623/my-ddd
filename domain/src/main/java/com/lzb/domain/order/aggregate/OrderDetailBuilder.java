package com.lzb.domain.order.aggregate;

import java.math.BigDecimal;

import com.lzb.component.helper.SpringHelper;
import com.lzb.component.id.IdGenerator;
import com.lzb.domain.order.enums.OrderStatus;
import lombok.RequiredArgsConstructor;

import org.springframework.beans.factory.config.ConfigurableBeanFactory;
import org.springframework.context.annotation.Lazy;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

/**
 * <br/>
 * Created on : 2023-09-09 13:57
 * @author lizebin
 */
@Component
@Scope(value = ConfigurableBeanFactory.SCOPE_PROTOTYPE)
@RequiredArgsConstructor(onConstructor = @__(@Lazy))
public class OrderDetailBuilder {

    private final IdGenerator idGenerator;

    ///////////////////////////////////////////////////////////////////////////
    // 属性值
    ///////////////////////////////////////////////////////////////////////////

    private long orderId;
    private int skuId;
    private BigDecimal price;
    private OrderStatus orderStatus;

    public static OrderDetailBuilder newInstance() {
        return SpringHelper.getBean(OrderDetailBuilder.class);
    }

    public OrderDetailBuilder orderId(long orderId) {
        this.orderId = orderId;
        return this;
    }

    public OrderDetailBuilder skuId(int skuId) {
        this.skuId = skuId;
        return this;
    }

    public OrderDetailBuilder price(BigDecimal price) {
        this.price = price;
        return this;
    }

    public OrderDetailBuilder orderStatus(OrderStatus orderStatus) {
        this.orderStatus = orderStatus;
        return this;
    }

    public OrderDetail build() {
        return new OrderDetail(idGenerator.id(), orderId, skuId, orderStatus, price);
    }

}
