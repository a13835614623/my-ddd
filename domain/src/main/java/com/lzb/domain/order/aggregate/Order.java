package com.lzb.domain.order.aggregate;

import java.math.BigDecimal;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.lzb.component.exception.BizException;
import com.lzb.domain.common.aggregate.BaseAggregate;
import com.lzb.domain.order.aggregate.builder.OrderAddressBuilder;
import com.lzb.domain.order.dto.LockStockDto;
import com.lzb.domain.order.enums.OrderStatus;
import com.lzb.domain.order.event.OrderCanceledEvent;
import com.lzb.domain.order.event.OrderPlacedEvent;
import com.lzb.domain.order.valobj.FullAddressLine;
import com.lzb.domain.order.valobj.FullName;
import lombok.Getter;
import lombok.NonNull;
import lombok.extern.slf4j.Slf4j;
import one.util.streamex.StreamEx;

/**
 * 订单聚合根<br/>
 * 实体所有写的操作，必须通过聚合根调用，所以改成包可见(default)
 * 实体所有读操作，可以改成公共(public)
 * Created on : 2022-06-11 14:41
 *
 * @author lizebin
 */
@Slf4j
@Getter
public class Order extends BaseAggregate<Order> {

    @NonNull
    private final OrderStatus orderStatus;

    @NonNull
    private final String currency;

    @NonNull
    private final BigDecimal exchangeRate;

    @NonNull
    private final BigDecimal totalShouldPay;

    @NonNull
    private final BigDecimal totalActualPay;

    @NonNull
    private final OrderAddress orderAddress;

    @NonNull
    private final OrderDetails orderDetails;


    /**
     * 仓储层使用，无需校验业务逻辑，直接构造
     * @param id
     * @param version
     * @param orderStatus
     * @param currency
     * @param exchangeRate
     * @param totalShouldPay
     * @param totalActualPay
     * @param orderAddress
     * @param orderDetails
     */
    public Order(long id,
            int version,
            @NonNull OrderStatus orderStatus,
            @NonNull String currency,
            @NonNull BigDecimal exchangeRate,
            @NonNull BigDecimal totalShouldPay,
            @NonNull BigDecimal totalActualPay,
            @NonNull OrderAddress orderAddress,
            @NonNull List<OrderDetail> orderDetails) {
        super(id);
        this.version = version;
        this.orderStatus = orderStatus;
        this.currency = currency;
        this.exchangeRate = exchangeRate;
        this.totalShouldPay = totalShouldPay;
        this.totalActualPay = totalActualPay;
        this.orderAddress = orderAddress;
        this.orderDetails = new OrderDetails(orderDetails);
    }


    public void updateAddress(String email,
            String phoneNumber, FullName fullName,
            FullAddressLine fullAddressLine, String country) {
        if (isShipped()) {
            throw new BizException("订单已发货，不能修改地址");
        }
        OrderAddress newOrderAddress = OrderAddressBuilder.newInstance(this.orderAddress)
                .addressLine1(fullAddressLine.getAddressLine1())
                .addressLine2(fullAddressLine.getAddressLine2())
                .country(country)
                .email(email)
                .firstName(fullName.getFirstName())
                .lastName(fullName.getLastName())
                .phoneNumber(phoneNumber).build();
        this.orderAddress = newOrderAddress;
    }

    /**
     * 更新库存结果
     * @param lockStock
     */
    public void updateLockStock(LockStockDto lockStock) {
        Map<Integer, List<OrderDetail>> skuId2OrderDetails = StreamEx.of(orderDetails.toStream()).groupingBy(OrderDetail::getSkuId);
        skuId2OrderDetails.keySet().forEach(skuId -> {
            List<OrderDetail> skuOrderDetails = skuId2OrderDetails.getOrDefault(skuId, Collections.emptyList());
            for (int i = 1; i <= skuOrderDetails.size(); i++) {
                skuOrderDetails.get(i - 1).updateLocked(i <= lockStock.getDetail(skuId).getLockedNum());
            }
        });
    }

    public void updateTotalActualPay(BigDecimal totalActualPay) {
        this.totalActualPay = totalActualPay;
    }

    /**
     * 取消订单
     */
    public void cancel() {
        orderStatus = OrderStatus.CANCELED;
        orderDetails.forEach(OrderDetail::cancel);
        addEvent(OrderCanceledEvent.create(id));
    }

    /**
     * 是否取消
     * @return
     */
    public boolean isCancel() {
        return OrderStatus.CANCELED == orderStatus;
    }

    /**
     * 生单
     */
    public void place() {
        addEvent(OrderPlacedEvent.create(this.id));
    }

    public boolean isShipped() {
        return false;
    }

    /**
     * 只更新姓名
     * @param fullName
     */
    public void updateFullName(FullName fullName) {
        this.orderAddress.updateFullName(fullName);
    }
}
