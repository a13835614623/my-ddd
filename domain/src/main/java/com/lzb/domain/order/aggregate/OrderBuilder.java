package com.lzb.domain.order.aggregate;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

import com.lzb.component.helper.SpringHelper;
import com.lzb.domain.common.aggregate.BaseBuilder;
import com.lzb.domain.order.enums.OrderStatus;
import com.lzb.domain.order.service.SkuValidator;
import com.lzb.domain.order.valobj.FullAddressLine;
import com.lzb.domain.order.valobj.FullName;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;

import org.springframework.beans.factory.config.ConfigurableBeanFactory;
import org.springframework.context.annotation.Lazy;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

/**
 * OrderBuilder用于app构建聚合根
 * Created on : 2023-09-07 22:52
 * @author mac
 */
@Component
@Scope(value = ConfigurableBeanFactory.SCOPE_PROTOTYPE)
@RequiredArgsConstructor(onConstructor = @__(@Lazy))
public class OrderBuilder extends BaseBuilder<Order> {

    private final SkuValidator skuValidator;

    ///////////////////////////////////////////////////////////////////////////
    // 属性值
    ///////////////////////////////////////////////////////////////////////////

    private long orderId;
    private String currency;
    private BigDecimal exchangeRate;
    private BigDecimal totalShouldPay;
    private BigDecimal totalActualPay;
    private String email;
    private String phoneNumber;
    private String firstName;
    private String lastName;
    private String addressLine1;
    private String addressLine2;
    private String country;
    private int version;
    private OrderStatus orderStatus;
    private final List<OrderDetail> innerOrderDetails = new ArrayList<>();

    // 聚合根内部值对象
    private FullName fullName;
    private FullAddressLine fullAddressLine;
    private OrderAddress orderAddress;
    private OrderDetails orderDetails;

    public static OrderBuilder newInstance() {
        return SpringHelper.getBean(OrderBuilder.class);
    }

    public OrderBuilder orderId(long orderId) {
        this.orderId = orderId;
        return this;
    }

    public OrderBuilder currency(@NonNull String currency) {
        this.currency = currency;
        return this;
    }

    public OrderBuilder exchangeRate(@NonNull BigDecimal exchangeRate) {
        this.exchangeRate = exchangeRate;
        return this;
    }

    public OrderBuilder totalShouldPay(@NonNull BigDecimal totalShouldPay) {
        this.totalShouldPay = totalShouldPay;
        return this;
    }

    public OrderBuilder totalActualPay(@NonNull BigDecimal totalActualPay) {
        this.totalActualPay = totalActualPay;
        return this;
    }

    public OrderBuilder email(@NonNull String email) {
        this.email = email;
        return this;
    }

    public OrderBuilder phoneNumber(@NonNull String phoneNumber) {
        this.phoneNumber = phoneNumber;
        return this;
    }

    public OrderBuilder firstName(@NonNull String firstName) {
        this.firstName = firstName;
        return this;
    }

    public OrderBuilder lastName(@NonNull String lastName) {
        this.lastName = lastName;
        return this;
    }

    public OrderBuilder addressLine1(@NonNull String addressLine1) {
        this.addressLine1 = addressLine1;
        return this;
    }

    public OrderBuilder addressLine2(@NonNull String addressLine2) {
        this.addressLine2 = addressLine2;
        return this;
    }

    public OrderBuilder country(@NonNull String country) {
        this.country = country;
        return this;
    }

    public OrderBuilder version(int version) {
        this.version = version;
        return this;
    }

    public OrderBuilder orderStatus(@NonNull OrderStatus orderStatus) {
        this.orderStatus = orderStatus;
        return this;
    }

    public OrderBuilder addOrderDetail(@NonNull OrderDetail orderDetail) {
        innerOrderDetails.add(orderDetail);
        return this;
    }


    @Override
    protected Order doBuild() {
        fullName = new FullName(firstName, lastName);
        fullAddressLine = new FullAddressLine(addressLine1, addressLine2);
        orderAddress = new OrderAddress(orderId, fullName, fullAddressLine, email, phoneNumber, country);
        orderDetails = new OrderDetails(this.innerOrderDetails);
        return new Order(orderId, version, orderStatus, currency, exchangeRate, totalShouldPay, totalActualPay, orderAddress, orderDetails);
    }

    @Override
    protected void validate(Order order) {
        skuValidator.assertAllOfSkuIsOnSale(order.getOrderDetails().getSkuIds());
    }

}
