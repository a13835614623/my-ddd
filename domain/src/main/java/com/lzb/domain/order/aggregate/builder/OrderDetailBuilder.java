package com.lzb.domain.order.aggregate.builder;

import java.math.BigDecimal;

import com.lzb.component.helper.SpringHelper;
import com.lzb.component.id.IdGenerator;
import com.lzb.domain.common.aggregate.BaseBuilder;
import com.lzb.domain.order.aggregate.OrderDetail;
import com.lzb.domain.order.enums.OrderStatus;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;

import org.springframework.beans.factory.config.ConfigurableBeanFactory;
import org.springframework.context.annotation.Lazy;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

/**
 * <br/>
 * Created on : 2023-09-09 13:57
 * @author lizebin
 */
@Component
@Scope(value = ConfigurableBeanFactory.SCOPE_PROTOTYPE)
@RequiredArgsConstructor(onConstructor = @__(@Lazy))
public class OrderDetailBuilder extends BaseBuilder<OrderDetail> {

    private final IdGenerator idGenerator;

    ///////////////////////////////////////////////////////////////////////////
    // 属性值
    ///////////////////////////////////////////////////////////////////////////

    private long id;
    private long orderId;
    private int skuId;
    private BigDecimal price;
    private OrderStatus orderStatus;
    private Boolean locked;

    public static OrderDetailBuilder newInstance() {
        return SpringHelper.getBean(OrderDetailBuilder.class);
    }

    public OrderDetailBuilder orderId(long orderId) {
        this.orderId = orderId;
        return this;
    }

    public OrderDetailBuilder skuId(int skuId) {
        this.skuId = skuId;
        return this;
    }

    public OrderDetailBuilder price(@NonNull BigDecimal price) {
        this.price = price;
        return this;
    }

    public OrderDetailBuilder orderStatus(@NonNull OrderStatus orderStatus) {
        this.orderStatus = orderStatus;
        return this;
    }

    public OrderDetailBuilder id(long id) {
        this.id = id;
        return this;
    }

    public OrderDetailBuilder locked(Boolean locked) {
        this.locked = locked;
        return this;
    }

    @Override
    protected OrderDetail doBuild() {
        return new OrderDetail(id, orderId, skuId, orderStatus, price, locked);
    }

}
