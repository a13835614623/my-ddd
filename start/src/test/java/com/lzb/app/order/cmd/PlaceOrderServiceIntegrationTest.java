package com.lzb.app.order.cmd;

import java.math.BigDecimal;
import java.util.List;

import com.lzb.BaseIntegrationTest;
import com.lzb.app.order.cmd.dto.PlaceOrderDetailReq;
import com.lzb.app.order.cmd.dto.PlaceOrderReq;
import com.lzb.component.id.IdGenerator;
import com.lzb.domain.order.aggregate.Order;
import com.lzb.domain.order.aggregate.OrderBuilder;
import com.lzb.domain.order.aggregate.OrderDetailBuilder;
import com.lzb.domain.order.enums.OrderStatus;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import org.springframework.beans.factory.annotation.Autowired;

/**
 * <br/>
 * Created on : 2023-09-09 14:23
 * @author lizebin
 */
class PlaceOrderServiceIntegrationTest extends BaseIntegrationTest {
    
    @Autowired
    private IdGenerator idGenerator;

    @Test
    @Disabled
    @DisplayName("测试OrderDetailBuilder")
    void should_place_order() {

        PlaceOrderReq req = new PlaceOrderReq("CNY", BigDecimal.ONE, BigDecimal.ONE, BigDecimal.ONE, "email",
                "phoneNumber", "firstName", "lastName", "addressLine1", "addressLine2", "country",
                List.of(new PlaceOrderDetailReq(1, BigDecimal.ONE)));

        long orderId = 1L;
        OrderBuilder orderBuilder = OrderBuilder.newInstance()
                .orderId(orderId)
                .currency(req.currency())
                .exchangeRate(req.exchangeRate())
                .totalShouldPay(req.totalShouldPay())
                .totalActualPay(req.totalActualPay())
                .email(req.email())
                .phoneNumber(req.phoneNumber())
                .firstName(req.firstName())
                .lastName(req.lastName())
                .addressLine1(req.addressLine1())
                .addressLine2(req.addressLine2())
                .country(req.country())
                .orderStatus(OrderStatus.WAIT_PAY);

        for (int i = 0; i < req.details().size(); i++) {
            PlaceOrderDetailReq detailReq = req.details().get(i);
            OrderDetailBuilder orderDetailBuilder = OrderDetailBuilder.newInstance()
                    .id(orderId + i + 1)
                    .orderId(orderId)
                    .skuId(detailReq.skuId())
                    .price(detailReq.price())
                    .orderStatus(OrderStatus.WAIT_PAY);
            orderBuilder.addDetailBuilder(orderDetailBuilder);
        }

        Order order = orderBuilder.build();
        assertJSON(order);
    }

}