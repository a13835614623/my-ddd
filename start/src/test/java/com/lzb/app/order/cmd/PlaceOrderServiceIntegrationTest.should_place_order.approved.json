{
  "currency" : "CNY",
  "events" : [ ],
  "exchangeRate" : 1,
  "id" : 1,
  "logs" : [ ],
  "orderAddress" : {
    "country" : "country",
    "email" : "email",
    "fullAddressLine" : {
      "addressLine1" : "addressLine1",
      "addressLine2" : "addressLine2"
    },
    "fullName" : {
      "firstName" : "firstName",
      "lastName" : "lastName"
    },
    "id" : 1,
    "phoneNumber" : "phoneNumber",
    "version" : 1
  },
  "orderDetails" : [ {
    "id" : 2,
    "orderId" : 1,
    "orderStatus" : "WAIT_PAY",
    "price" : 1,
    "skuId" : 1,
    "version" : 1
  } ],
  "orderStatus" : "WAIT_PAY",
  "totalActualPay" : 1,
  "totalShouldPay" : 1,
  "version" : 0
}